[CI] RepositoryVerifyIntegrityIT testCorruption failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #3995 / rocky-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/vdbicx35akwn6)
- [elasticsearch-periodic-platform-support #3995 / amazonlinux-2023_platform-support-aws](https://gradle-enterprise.elastic.co/s/bmpyl2547zefe)
- [elasticsearch-pull-request #30817 / part-2](https://gradle-enterprise.elastic.co/s/n5guiktkdftsq)
- [elasticsearch-pull-request #30797 / part-2](https://gradle-enterprise.elastic.co/s/zyeurze3sk4qs)
- [elasticsearch-intake #9559 / part2](https://gradle-enterprise.elastic.co/s/46dvnpy6zmjzq)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:snapshot-repo-test-kit:internalClusterTest" --tests "org.elasticsearch.repositories.blobstore.testkit.integrity.RepositoryVerifyIntegrityIT.testCorruption" -Dtests.seed=4C617F002EA1A2D6 -Dtests.locale=be-BY -Dtests.timezone=America/Danmarkshavn -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testCorruption'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.repositories.blobstore.testkit.integrity.RepositoryVerifyIntegrityIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: not an empty collection
     but: was <[]>
```

**Issue Reasons:**
- [main] 5 failures in test testCorruption (12.8% fail rate in 39 executions)
- [main] 2 failures in step part-2 (10.5% fail rate in 19 executions)
- [main] 2 failures in pipeline elasticsearch-pull-request (11.1% fail rate in 18 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -208,9 +208,6 @@ tests:
 - class: org.elasticsearch.smoketest.DocsClientYamlTestSuiteIT
   method: test {yaml=reference/cluster/stats/line_1450}
   issue: https://github.com/elastic/elasticsearch/issues/112732
-- class: org.elasticsearch.repositories.blobstore.testkit.integrity.RepositoryVerifyIntegrityIT
-  method: testCorruption
-  issue: https://github.com/elastic/elasticsearch/issues/112769
 
 # Examples:
 #

--- a/x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/integrity/RepositoryVerifyIntegrityIT.java
+++ b/x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/integrity/RepositoryVerifyIntegrityIT.java
@@ -344,16 +344,21 @@ public class RepositoryVerifyIntegrityIT extends AbstractSnapshotIntegTestCase {
                 ? equalTo(testContext.indexNames().size())
                 : lessThan(testContext.indexNames().size())
         );
-        assertThat(anomalies, not(empty()));
+        // Missing shard generation file is automatically repaired based on the shard snapshot files.
+        // See also BlobStoreRepository#buildBlobStoreIndexShardSnapshots
+        assertThat(anomalies, corruptedFileType == RepositoryFileType.SHARD_GENERATION ? empty() : not(empty()));
         assertThat(responseObjectPath.evaluate("results.total_anomalies"), greaterThanOrEqualTo(anomalies.size()));
-        assertEquals("fail", responseObjectPath.evaluate("results.result"));
+        assertEquals(
+            corruptedFileType == RepositoryFileType.SHARD_GENERATION ? "pass" : "fail",
+            responseObjectPath.evaluate("results.result")
+        );
 
         // remove permitted/expected anomalies to verify that no unexpected ones were seen
         switch (corruptedFileType) {
             case SNAPSHOT_INFO -> anomalies.remove("failed to load snapshot info");
             case GLOBAL_METADATA -> anomalies.remove("failed to load global metadata");
             case INDEX_METADATA -> anomalies.remove("failed to load index metadata");
-            case SHARD_GENERATION -> anomalies.remove("failed to load shard generation");
+            case SHARD_GENERATION -> {}
             case SHARD_SNAPSHOT_INFO -> anomalies.remove("failed to load shard snapshot");
             case SHARD_DATA -> {
                 anomalies.remove("missing blob");

--- a/x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/integrity/RepositoryVerifyIntegrityIT.java
+++ b/x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/integrity/RepositoryVerifyIntegrityIT.java
@@ -358,8 +358,6 @@ public class RepositoryVerifyIntegrityIT extends AbstractSnapshotIntegTestCase {
             case SNAPSHOT_INFO -> anomalies.remove("failed to load snapshot info");
             case GLOBAL_METADATA -> anomalies.remove("failed to load global metadata");
             case INDEX_METADATA -> anomalies.remove("failed to load index metadata");
-            case SHARD_GENERATION -> {
-            }
             case SHARD_SNAPSHOT_INFO -> anomalies.remove("failed to load shard snapshot");
             case SHARD_DATA -> {
                 anomalies.remove("missing blob");

--- a/x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/integrity/RepositoryVerifyIntegrityIT.java
+++ b/x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/integrity/RepositoryVerifyIntegrityIT.java
@@ -358,7 +358,8 @@ public class RepositoryVerifyIntegrityIT extends AbstractSnapshotIntegTestCase {
             case SNAPSHOT_INFO -> anomalies.remove("failed to load snapshot info");
             case GLOBAL_METADATA -> anomalies.remove("failed to load global metadata");
             case INDEX_METADATA -> anomalies.remove("failed to load index metadata");
-            case SHARD_GENERATION -> {}
+            case SHARD_GENERATION -> {
+            }
             case SHARD_SNAPSHOT_INFO -> anomalies.remove("failed to load shard snapshot");
             case SHARD_DATA -> {
                 anomalies.remove("missing blob");

--- a/x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/integrity/RepositoryVerifyIntegrityIT.java
+++ b/x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/integrity/RepositoryVerifyIntegrityIT.java
@@ -346,18 +346,21 @@ public class RepositoryVerifyIntegrityIT extends AbstractSnapshotIntegTestCase {
         );
         // Missing shard generation file is automatically repaired based on the shard snapshot files.
         // See also BlobStoreRepository#buildBlobStoreIndexShardSnapshots
-        assertThat(anomalies, corruptedFileType == RepositoryFileType.SHARD_GENERATION ? empty() : not(empty()));
+        final boolean deletedShardGen = corruptedFileType == RepositoryFileType.SHARD_GENERATION && Files.exists(corruptedFile) == false;
+        assertThat(anomalies, deletedShardGen ? empty() : not(empty()));
         assertThat(responseObjectPath.evaluate("results.total_anomalies"), greaterThanOrEqualTo(anomalies.size()));
-        assertEquals(
-            corruptedFileType == RepositoryFileType.SHARD_GENERATION ? "pass" : "fail",
-            responseObjectPath.evaluate("results.result")
-        );
+        assertEquals(deletedShardGen ? "pass" : "fail", responseObjectPath.evaluate("results.result"));
 
         // remove permitted/expected anomalies to verify that no unexpected ones were seen
         switch (corruptedFileType) {
             case SNAPSHOT_INFO -> anomalies.remove("failed to load snapshot info");
             case GLOBAL_METADATA -> anomalies.remove("failed to load global metadata");
             case INDEX_METADATA -> anomalies.remove("failed to load index metadata");
+            case SHARD_GENERATION -> {
+                if (deletedShardGen == false) {
+                    anomalies.remove("failed to load shard generation");
+                }
+            }
             case SHARD_SNAPSHOT_INFO -> anomalies.remove("failed to load shard snapshot");
             case SHARD_DATA -> {
                 anomalies.remove("missing blob");

